src/main/java/emu/grasscutter/game/managers/StaminaManager/StaminaManager.java

```diff
@@ -29,100 +29,174 @@ public class StaminaManager {
+                MotionState.MOTION_POWERED_FLY
         )));
 
         MotionStatesCategorized.put("RUN", new HashSet<>(Arrays.asList(
-            MotionState.MOTION_DASH,
-            MotionState.MOTION_DANGER_DASH,
-            MotionState.MOTION_DASH_BEFORE_SHAKE,
-            MotionState.MOTION_RUN,
-            MotionState.MOTION_DANGER_RUN,
-            MotionState.MOTION_WALK,
-            MotionState.MOTION_DANGER_WALK
+                MotionState.MOTION_DASH,
+                MotionState.MOTION_DANGER_DASH,
+                MotionState.MOTION_DASH_BEFORE_SHAKE,
+                MotionState.MOTION_RUN,
+                MotionState.MOTION_DANGER_RUN,
+                MotionState.MOTION_WALK,
+                MotionState.MOTION_DANGER_WALK
         )));
 
         MotionStatesCategorized.put("FIGHT", new HashSet<>(Arrays.asList(
-            MotionState.MOTION_FIGHT
+                MotionState.MOTION_FIGHT
         )));
+
+        MotionStatesCategorized.put("SKIFF", new HashSet<>(Arrays.asList(
+                MotionState.MOTION_SKIFF_BOARDING,
+                MotionState.MOTION_SKIFF_NORMAL,
+                MotionState.MOTION_SKIFF_DASH,
+                MotionState.MOTION_SKIFF_POWERED_DASH
+        )));
+    }
+
+    // Listeners
+
+    public boolean registerBeforeUpdateStaminaListener(String listenerName, BeforeUpdateStaminaListener listener) {
+        if (beforeUpdateStaminaListeners.containsKey(listenerName)) {
+            return false;
+        }
+        beforeUpdateStaminaListeners.put(listenerName, listener);
+        return true;
+    }
+
+    public boolean unregisterBeforeUpdateStaminaListener(String listenerName) {
+        if (!beforeUpdateStaminaListeners.containsKey(listenerName)) {
+            return false;
+        }
+        beforeUpdateStaminaListeners.remove(listenerName);
+        return true;
+    }
+
+    public boolean registerAfterUpdateStaminaListener(String listenerName, AfterUpdateStaminaListener listener) {
+        if (afterUpdateStaminaListeners.containsKey(listenerName)) {
+            return false;
+        }
+        afterUpdateStaminaListeners.put(listenerName, listener);
+        return true;
+    }
+
+    public boolean unregisterAfterUpdateStaminaListener(String listenerName) {
+        if (!afterUpdateStaminaListeners.containsKey(listenerName)) {
+            return false;
+        }
+        afterUpdateStaminaListeners.remove(listenerName);
+        return true;
     }
 
     private boolean isPlayerMoving() {
         float diffX = currentCoordinates.getX() - previousCoordinates.getX();
         float diffY = currentCoordinates.getY() - previousCoordinates.getY();
         float diffZ = currentCoordinates.getZ() - previousCoordinates.getZ();
-        Grasscutter.getLogger().debug("isPlayerMoving: " + previousCoordinates + ", " + currentCoordinates +
+        Grasscutter.getLogger().trace("isPlayerMoving: " + previousCoordinates + ", " + currentCoordinates +

```

Comment:
Also needs indication on where it came from, I propose `StaminaManager::isPlayerMoving`. `isPlayerMoving` alone is pretty generic name that I don't doubt will be reused.